Store client and tagger tags separately

vickenty opened the pull request:
### What does this PR do?

Store client tags and origin detection tags separately in the dogstatsd context resolver.

### Motivation

Reduce memory usage when agent handles traffic from containers with origin detection enabled.

### Additional Notes

This PR contains only the new feature; a cleanup refactoring will follow in a separate PR.

This patch is based on observation that tagger tags usually have lower cardinality than metric tags. If stored separately, origin detection tags can be shared between higher number of contexts, reducing number of tags in the per-context slice (which can also be shared better, if the same set of tags happens to come from different origins).

Benefits depends greatly on specific traffic pattern handled by the agent, namely the fraction of tags supplied by the tagger per metric, as well as total number of contexts. 

### Possible Drawbacks / Trade-offs

Tags deduplication code became more complicated, and new code uses ~2% more CPU than previously; even with origin detection disabled.

### Describe how to test/QA your changes

Deploy the agent on docker with telemetry, payload logging and origin detection:

```sh
docker run --rm --name qa \
       -v /var/run/docker.sock:/var/run/docker.sock:ro \
       -v /proc/:/host/proc/:ro \
       -v /sys/fs/cgroup/:/host/sys/fs/cgroup:ro \
       -v /var/run/datadog:/var/run/datadog \
       -e DD_API_KEY=none \
       -e DD_DOGSTATSD_SOCKET=/var/run/datadog/dsd.sock \
       -e DD_DOGSTATSD_ORIGIN_DETECTION=true \
       -e DD_DOGSTATSD_TAG_CARDINALITY=high \
       -e DD_LOG_LEVEL=debug \
       -e DD_LOG_PAYLOADS=true \
       -e DD_TELEMETRY_ENABLED=true \
       --pid host \
       datadog/agent:7-rc
```

With the agent running, start a mock client to send some metrics via uds, and:
- [ ] Check that client tags (and only client tags) are sent to the intake on k8s when origin detection is disabled.
- [ ] Check that tags sent are correct (match the metric they were attached to in the client).
- [ ] Check that both client and origin detection tags are sent to the intake, tags are attached to correct metrics.
- [ ] Stop the client; check after 5 minutes that the number of tagsets tracked by the aggregator_tags_store goes down;
  `docker exec qa curl -s http://localhost:5000/telemetry | grep aggregator_tag_store__entries`

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).

Note: Adding GitHub labels is only possible for contributors with write access.
-->

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] The appropriate `team/..` label has been applied, if known.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

ogaca-dd committed:
Composite tags (#10460)

* Introduce `NewCompositeTags`

vickenty committed:
Add benchmark to compare two hasing methods

GenerateWithTags vs GenerateWithTags2

vagrant@bullseye:/go/src/github.com/DataDog/datadog-agent/pkg/aggregator/ckey$ benchcmp -best ~/bench-tags-1 ~/bench-tags-2
benchcmp is deprecated in favor of benchstat: https://pkg.go.dev/golang.org/x/perf/cmd/benchstat
benchmark                               old ns/op     new ns/op     delta
BenchmarkKeyGeneration2/1-tags-8        17.3          19.8          +14.26%
BenchmarkKeyGeneration2/2-tags-8        19.8          21.9          +10.57%
BenchmarkKeyGeneration2/4-tags-8        30.6          33.1          +8.45%
BenchmarkKeyGeneration2/8-tags-8        33.7          40.7          +20.82%
BenchmarkKeyGeneration2/16-tags-8       54.9          64.4          +17.19%
BenchmarkKeyGeneration2/32-tags-8       83.0          100           +21.06%
BenchmarkKeyGeneration2/64-tags-8       151           184           +21.33%
BenchmarkKeyGeneration2/128-tags-8      374           352           -5.73%
BenchmarkKeyGeneration2/256-tags-8      555           658           +18.42%
BenchmarkKeyGeneration2/512-tags-8      10212         10625         +4.04%
BenchmarkKeyGeneration2/1024-tags-8     57443         57980         +0.93%
BenchmarkKeyGeneration2/2048-tags-8     129906        132638        +2.10%

vickenty committed:
Update otlp/serilizerexporter to use composite tags

Tags enrichment was added on main and it needs to be updated to use
CompositeTags.

This patch adds CombineWithSlice to modify CompositeTags inplace, to
avoid mistakes like the one I did today:

    c.sketches[i].Tags = CombineCompositeTagsWithSlice(c.series[i].Tags, ..)

vickenty committed:
Merge remote-tracking branch 'origin/main' into dogstatsd-tags

vickenty committed:
Add HashGenerator.Hash2

A new method to deduplicate tags and compute hashes from two builders
at the same time.

vickenty committed:
Store metric and tagger tags separately

This patch is based on observation that tagger tags usually have lower
cardinality than metric tags. If stored separately, tagger tags can be
shared between higher number of contexts, reducing number of tags in
the per-context slice (which can also be shared better, if the same
tagset happens to come from different origins).

Use separate accumulators for tags that came with the metric sample
and for tags from the tagger (origin detection). Use Hash2 to
deduplicate them together and use hashes to store them separately in
the store. Leverage CompositeTags to pass the two tags slices
separately all the way to the serializer without having to allocate
unified slice most of the time.

vickenty committed:
Move hashing out of hash_generator

This has slightly neagtive impact on performance, but makes complex
deduplication code a bit simpler and more focused.

benchmark                               old ns/op     new ns/op     delta
BenchmarkKeyGeneration2/1-tags-8        19.8          21.4          +8.14%
BenchmarkKeyGeneration2/2-tags-8        21.9          24.6          +12.72%
BenchmarkKeyGeneration2/4-tags-8        33.1          38.5          +16.33%
BenchmarkKeyGeneration2/8-tags-8        40.7          42.8          +4.98%
BenchmarkKeyGeneration2/16-tags-8       64.4          66.5          +3.29%
BenchmarkKeyGeneration2/32-tags-8       100           106           +5.37%
BenchmarkKeyGeneration2/64-tags-8       184           190           +3.27%
BenchmarkKeyGeneration2/128-tags-8      352           359           +1.90%
BenchmarkKeyGeneration2/256-tags-8      658           666           +1.19%
BenchmarkKeyGeneration2/512-tags-8      10625         10291         -3.14%
BenchmarkKeyGeneration2/1024-tags-8     57980         57724         -0.44%
BenchmarkKeyGeneration2/2048-tags-8     132638        130137        -1.89%

vickenty committed:
gofmt

djmitche commented on pkg/metrics/histogram_bucket.go:
```diff
@@ -33,9 +33,9 @@ func (m *HistogramBucket) GetHost() string {
 
 // GetTags returns the bucket tags.
-func (m *HistogramBucket) GetTags(tb *tagset.HashingTagsAccumulator) {
+func (m *HistogramBucket) GetTags(taggerBuffer, metricBuffer *tagset.HashingTagsAccumulator) {
```
why pass taggerBuffer here?  It's shared, so it would be pretty bad if this function modified it, right?

vickenty replied:
It's part of the interface. If HistogramBuckets supported origin detection, this function would be required to write tags from the origin detection to this accumulator. The contract is indeed to modify the accumulator, so it shouldn't be shared.

remeh replied:
If I understand correctly, the `GetTags(*tagset.HashingTagsAccumulator, *tagset.HashingTagsAccumulator)` doesn't guard against an implementation modifying the `taggerBuffer` which should not be modified, right?

vickenty replied:
No, it is very much should to be modified. It comes in empty and the implementation should fill it with tags from origin detection. (Doesn't matter too much which one even, except we keep tags in the first one when there are duplicates).

djmitche reviewed (changes requested):
This is a very complex change, and only sparsely documented.  Even looking at this PR specifically, it's hard to see how it all fits together, or discover the hidden dependencies.  I can't imagine trying to figure this out, or be confident that I wasn't violating some invariant, if I were trying to modify this code later.  And I think we _will_ want to modify this code later.

In general the benchmarks indicate that this is a good idea, but I think it needs a good deal more work on maintainability.  Much of that may be extensive documentation, perhaps with the addition of some runtime verification where that's not in the hot path.  It might also be helpful to have some test-only functionality, such as to assert that a HTA is unique, that can be used in tests in other modules.

vickenty committed:
Update pkg/tagset/composite_tags.go

Co-authored-by: Dustin J. Mitchell <dustin.mitchell@datadoghq.com>

vickenty committed:
Update pkg/tagset/composite_tags.go

Co-authored-by: Dustin J. Mitchell <dustin.mitchell@datadoghq.com>

vickenty commented on pkg/tagset/hash_generator.go:
```diff
@@ -135,3 +135,110 @@ func (g *HashGenerator) Hash(tb *HashingTagsAccumulator) uint64 {
+// tag each tag is present once in either l or r, but not both at the same time.
+//
+// First, duplicates are removed from l. Then duplicates are removed from r, including any tags that
+// are already present in l. Can move tags from one accumulator to another.
```
No, the direction does not matter. There are only two instances of HTA used with this function. We copy tags to them, compute the hashes and deduplicate, and then copy out if the tags are new and we need to store them. Accumulators are then cleared and re-used for another metric. Which accumulator ends up holding which tags affects only our ability to share them effectively between contexts (so we want to keep lower cardinality tags from origin detection separate from potentially higher cardinality tags sent by the client), but otherwise we don't need to make any distinction between the two.

vickenty commented on pkg/metrics/metric_sample.go:
```diff
@@ -95,9 +95,9 @@ func (m *MetricSample) GetHost() string {
-	tagger.EnrichTags(tb, m.OriginFromUDS, m.OriginFromClient, m.Cardinality)
+func (m *MetricSample) GetTags(taggerBuffer, metricBuffer *tagset.HashingTagsAccumulator) {
+	metricBuffer.Append(m.Tags...)
+	tagger.EnrichTags(taggerBuffer, m.OriginFromUDS, m.OriginFromClient, m.Cardinality)
```
No, we don't share the accumulators between contexts.

ogaca-dd commented on pkg/tagset/hash_generator.go:
```diff
@@ -135,3 +135,121 @@ func (g *HashGenerator) Hash(tb *HashingTagsAccumulator) uint64 {
+//
+// First, duplicates are removed from l. Then duplicates are removed from r, including any tags that
+// are already present in l. Can move tags from one accumulator to another.
+func (g *HashGenerator) Hash2(l *HashingTagsAccumulator, r *HashingTagsAccumulator) (uint64, uint64) {
```
My initial thoughts on this was to always keep orchestrator tags unchanged (as they don't change often) and update only metrics tags by removing tags that are both in metrics and orchestrator. Doing so allow orchestrator tags to be reused very often. It would be also possible to cache a `map[string]struct{}` containing the hashes of the orchestrator tags to make the implementation of this function efficient.
Why do you move tags between accumulators?

vickenty replied:
Orchestrator tags may come from several different sources inside the tagger (two different origins, and separate "orchestrator" entity tags); and thus may, in theory, contain duplicates. Tags come with precomputed hashes from the tagger, so all we do is shift some integers around as we fill the internal hashset, the deduplication part is not expensive. We don't use a `map[string]struct{}` currently, as we need individual numeric hash values to construct the context key, and want to hash each tag once.

Moving only happens when the combined size of the two accumulators exceeds the size of our pre-allocated hashset. In this case the fallback behavior is to copy all tags to a single accumulator and `SortUniq`. This isn't expected to happen in practice as the size of the hashset is sufficiently high; we just didn't want to have any sort of artificial limitation on number of tags the agent can handle.

vickenty committed:
Add comment for GetTags.

Describe expected behavior and meaning of the parameters.

vickenty committed:
Add comments about implementation.

vickenty committed:
Expand comments on CompositeTags

Highlight read-only nature of the type, add an implementation
reminder, and annotate methods that take slices that slices will be aliased.

remeh commented on pkg/tagset/composite_tags.go:
```diff
@@ -0,0 +1,131 @@
+
+// UnmarshalJSON implements the json.Unmarshaler interface.
+// UnmarshalJSON receiver need to be a pointer to modify `t`.
+func (t *CompositeTags) UnmarshalJSON(b []byte) error {
```
Do you know where it is used?

vickenty replied:
I believe @ogaca-dd just added it for completeness.

vickenty committed:
Merge remote-tracking branch 'origin/main' into vickenty/tkey-2-clean

vickenty committed:
Add more tests

vickenty committed:
Verify hash values for deduplicated tags

vickenty committed:
Add more mid-sized test cases

vickenty commented on pkg/tagset/composite_tags.go:
```diff
@@ -0,0 +1,131 @@
+
+// CompositeTags stores two tag sets and provides methods to manipulate them easily
+// CompositeTags is designed to be used for metric tags created by the aggregator
+// (Context, Serie, SketchSeries, ...)
```
I added some comments to the type, its constructors and mutators that should cover potential aliasing issues.

vickenty commented on pkg/metrics/metric_sample.go:
```diff
@@ -63,7 +63,7 @@ func (m MetricType) String() string {
 	GetName() string
 	GetHost() string
-	GetTags(*tagset.HashingTagsAccumulator)
+	GetTags(*tagset.HashingTagsAccumulator, *tagset.HashingTagsAccumulator)
```
I added comment for GetTags.

vickenty commented on pkg/tagset/hash_generator.go:
```diff
@@ -135,3 +135,110 @@ func (g *HashGenerator) Hash(tb *HashingTagsAccumulator) uint64 {
+			ibase = int16(ntags)
+			tb.Truncate(ntags)
+		}
+	} else {
```
Done.

vickenty commented on pkg/tagset/composite_tags.go:
```diff
@@ -0,0 +1,131 @@
+// UnsafeToReadOnlySliceString creates a new slice containing all tags.
+// The caller of this method must ensure that the slice is never mutated.
+// Should be used only for performance reasons.
+func (t CompositeTags) UnsafeToReadOnlySliceString() []string {
```
I think we should be able to get rid of `Unsafe` altogether in the follow-up PR: it will make a copy most of the time anyway (to join two tags slices from the aggregator), so might as well be safe and always copy.

vickenty commented on pkg/tagset/composite_tags.go:
```diff
@@ -0,0 +1,131 @@
+}
+
+// CombineWithSlice adds tags to the composite tags. Consumes the slice.
+func (t *CompositeTags) CombineWithSlice(tags []string) {
```
Good point. I re-wrote the comments a bit, but safety notices are now duplicated across all mutating methods.

vickenty commented on pkg/tagset/hash_generator_test.go:
```diff
@@ -113,3 +115,85 @@ func genTags(count int, div int) ([]string, []string) {
+		{"empty r small", "foo", "", "foo", ""},
+		{"small-1", "foo,foo,bar", "ook", "foo,bar", "ook"},
+		{"small-2", "foo,bar", "foo,ook", "foo,bar", "ook"},
+		{"small-3", "foo", "ook,foo,eek", "foo", "ook,eek"},
```
Added. "small-6" is technically not small, since total number of tags is larger than 4 (this was originally testing only the nested-loop variant of deduplication used for small tagsets). I renamed it to "mid-1".

vickenty committed:
Add reproducable test for GenerateWithTags2

vickenty commented on pkg/aggregator/ckey/key.go:
```diff
@@ -61,6 +61,18 @@ func (g *KeyGenerator) GenerateWithTags(name, hostname string, tagsBuf *tagset.H
+//
+// Tags from l, r are combined to produce the key and deduplicated, but most of the time left in
+// their respective buffers.
+func (g *KeyGenerator) GenerateWithTags2(name, hostname string, l, r *tagset.HashingTagsAccumulator) (ContextKey, TagsKey, TagsKey) {
```
Good point, added.

djmitche commented on pkg/tagset/composite_tags.go:
```diff
@@ -10,15 +10,20 @@ import (
-// CompositeTags is designed to be used for metric tags created by the aggregator
-// (Context, Serie, SketchSeries, ...)
+// CompositeTags stores read-only views of two tag sets and provides methods to iterate them easily.
+
```
Add `//`

vickenty committed:
Call Shrink for time sampler tag stores

vickenty committed:
Fix comment formatting

remeh reviewed (approved):
Great work!

vickenty committed:
Merge remote-tracking branch 'origin/main' into vickenty/tkey-2-clean
